online_party_system/party_match.go

```diff
@@ -93,24 +93,86 @@ type PartyMatchState struct {
 	creator           string           // Party creator user ID, used to allow the user to automatically join when they create.
 	initialEmptyTicks int              // Number of ticks the party has been empty on creation, to ensure parties are cleaned up if their creator never joins.
 }
 
 type PartyConfig struct {
-	MaxSize        			int           			// Maximum number of members allowed in party
-	InviteDuration 			time.Duration 			// Duration an invite should last before expiring. default 0 = no expiration
-	MatchJoinMetadataFilter MatchJoinMetadataFilter	// 
-	MatchEndHook            MatchEndHook			// Called when the party match is over
-	MatchInitHook           MatchInitHook			// Called when the party match is initialized
-	MatchJoinAttemptHook    MatchJoinAttemptHook	// Called when a user attempts to join the party
-	MatchKickHook 			MatchKickHook			// Called when a party member is kicked from the party
-	MatchLeaveHook 			MatchLeaveHook			// Called when a party member leaves the party
+	MaxSize                 int                     // Maximum number of members allowed in party
+	InviteDuration          time.Duration           // Duration an invite should last before expiring. default 0 = no expiration
+	MatchJoinMetadataFilter MatchJoinMetadataFilter //
+	MatchEndHook            MatchEndHook            // Called when the party match is over
+	MatchInitHook           MatchInitHook           // Called when the party match is initialized
+	MatchJoinAttemptHook    MatchJoinAttemptHook    // Called when a user attempts to join the party
+	MatchKickHook           MatchKickHook           // Called when a party member is kicked from the party
+	MatchLeaveHook          MatchLeaveHook          // Called when a party member leaves the party
 }
 
 type PartyMatch struct {
-	config                  PartyConfig
+	config PartyConfig
+}
+
+func addUsersBlockedFriendsToBlockedSet(ctx context.Context, logger runtime.Logger, nk runtime.NakamaModule, userID string, blockedUsers map[string]struct{}) {
+	friendLimit := 100 // FriendsList will fail if we ask for > 100
+	friendState := 3   // 3 is Friend_BLOCKED
+	cursor := ""       // no cursor
+
+	friendsPtrArray, cursor, err := nk.FriendsList(ctx, userID, friendLimit, &friendState, cursor)
+	if err != nil {
+		logger.Error("Error getting FriendsList: %v", err)
+		return
+	}
+
+	for _, friendPtr := range friendsPtrArray {
+		if friendPtr != nil {
+			friend := *friendPtr
+			user := friend.User
+			if user != nil {
+				userID := user.Id
+				blockedUsers[userID] = struct{}{}
+			}
+		}
+	}
+}
+
+func getUsersBlockedFriends(ctx context.Context, logger runtime.Logger, nk runtime.NakamaModule, userID string) map[string]struct{} {
+	blockedSet := make(map[string]struct{})
+	addUsersBlockedFriendsToBlockedSet(ctx, logger, nk, userID, blockedSet)
+	return blockedSet
+}
+
+func rebuildBlockedUsersIfNeeded(ctx context.Context, logger runtime.Logger, nk runtime.NakamaModule, s *PartyMatchState) {
```

Comment:
If I'm reading the code right it seems this is mostly needed because when a user joins/leaves we don't know which of the users in the blocked set are there because of each party member. Perhaps instead we should keep track of who exactly each party member is blocking so we can drop them without going back to the database.

Would be a little more work when writing users into the blocked set but saves database ops I think?

Reply:
Yeah, no problems